chapters/chap03-sec03.tex

```diff
@@ -63,12 +63,13 @@ \subsection{Instantiating the packing functions}\label{instantiation:statement:p
     \caption{Packing of string $A$ (see~\cref{instantiation:eg:packing})}\label{instantiation:fig:packingA}
 \end{figure}
 
-We denote by $\resBitsBLen$ the total number of bits which could not fully fill a field element, i.e. the weighted sum of the $\gamma_x$,
+We now consider strategies to pack all primary inputs that are binary strings. A naive approach is to encode each binary string $x$ as $\alpha_x$ field elements. In general, this results in significant waste (and consequently more field elements than necessary), especially when the number of residual bits is small compared to $\bnFieldBitCap$ (see~\cref{instantiation:fig:packing-multi-strings}). An alternative strategy could be to concatenate all binary strings into a single string $y$ and pack this string into $\alpha_y$ field elements. While this approach minimizes the set of unused bits, it significantly increases the complexity of the unpacking operation.
```

Comment:
`increases the complexity of the unpacking operation` -> I would briefly expand on that to provide more background to the reader who may not realize/remember that this operation is done on-chain (on the EVM) hence why we'd rather avoid adding complexity. In fact, this operation is annoying is carried out on chain because we need to increase "bit shifts" as we decode the elements (the field capacity/EVM word length mismatch "accumulates" along the decoding), and doing bit-wise arithmetic isn't that cheap on-chain.

Reply:
I've expanded this a bit. This extra complexity could be handled as complex logic to compute dynamic shifts and masks (higher gas cost) or as many hard-coded unpack operations (minimal, if any, extra gas cost but more code and less maintainable).  I've tried to capture this.

Reply:
Yes, thanks.

When you say:
> This significantly increases the complexity of the unpacking operation that must be performed on-chain, resulting in a higher gas cost (due to extra logic) or more contract code (if each unpack operation is hard-coded)

Isn't 2 (i.e. `more contract code (if each unpack operation is hard-coded)`) already encapsulated in 1 (i.e. `resulting in a higher gas cost (due to extra logic)`)? Having more code => more OPCODES to pay gas for (and more stack operations) on-chain

Reply:
I may be missing something, but it's not clear to me that the gas cost would have to be significantly greater.  I think it's a bit implementation-dependent, hence I origianally just called it "complexity", rather than mentioning on-chain "cost".

Our (current) code does 2 shifts/masks per unpack.  Using the single string approach this is pretty much what would be required per unpack (except in some extreme cases where a string spans 3 field elements).  Unless I'm missing something, the bulk of any extra gas cost would come from dynamically computing *how much* to shift /mask with. So I didn't want to call this a purely "on-chain" cost, since technically these values could be hard-coded.

The second point tries to illustrate that. If the shifts / masks are hard-coded, the *complexity* is then in the source code (which, admittedly, would be pretty terrible) but at runtime the number of shifts / masks performed would not be so different from the current implementation (2 shifts / masks per unpack).  There would be more instructions in the contract (because each unpack would be "unrolled") so deployment might be marginally more expensive, but the number of instructions executed (and thereby the gas cost per tx) is not necessarily greater.

... if that makes sense.  Happy to rephrase, though.  I didn't want to get into too much detail in the text, but didn't want to imply that this would always result in much more expensive transactions.

Reply:
Makes sense, thanks